scripts/api_report.py

```diff
@@ -0,0 +1,195 @@
+    for folder_name, _, file_list in os.walk(examples_folder):
+        for file_name in [f for f in file_list if f.lower() == METADATA_FILENAME]:
+            file_path = os.path.join(folder_name, file_name)
+            print(f"Found metadata: {file_path}.")
+            read_metadata(file_path, examples)
+    return examples
+
+
+def read_metadata(file_path, examples):
+    """
+    Read the specified file and append its contents into the specified list
+    of dictionaries.
+
+    :type file_path: string
+    :param file_path: A metadata file that contains example metadata in yaml format.
+    :type examples: list
+    :param examples: A list of example dictionaries. Examples read from the metadata
+                     are appended to this list.
+    """
+    with open(file_path, 'r') as meta_stream:
+        try:
+            meta_docs = yaml.safe_load_all(meta_stream)
+            for example_meta in meta_docs:
+                example_meta['metadata_path'] = file_path
+                examples.append(example_meta)
+        except ScannerError as err:
+            print(f"Yaml parser error in {file_path}, skipping.")
+            print(err)
+
+
+def write_report(examples, report_path=None):
+    """
+    Writes a report of APIs covered by the list of examples. The report includes
+    the count of unique APIs, total API and example counts, and the full list of
+    examples in CS format.
+
+    :type examples: list
+    :param examples: A list of example dictionaries.
+    :type report_path: string
+    :param report_path: The output file to write the report. If this file exists,
+                        it is overwritten. If no file is specified, the report
+                        is written to sys.stdout.
+    :rtype: int
+    :return: The count of unique APIs covered by the examples list.
+    """
+    lines = ["Created,File,Language,Service,Operation"]
+    unique_apis = set()
+
+    for example in examples:
+        try:
+            created = example['created']
+            for file in example['files']:
+                metadata_folder = os.path.split(example['metadata_path'])[0]
+                metadata_url = request.pathname2url(metadata_folder)
+                base_url = urljoin(GITHUB_URL, metadata_url) + '/'
+                file_url = request.pathname2url(file['path'])
+                file_url = urljoin(base_url, file_url)
+                ext = os.path.splitext(file_url)[1].lstrip('.')
+                language = EXT_LOOKUP[ext]
+                for api in file['apis']:
+                    service = api['service']
+                    for operation in api['operations']:
+                        unique_apis.add((language, service, operation))
+                        lines.append(
+                            ','.join([str(created), file_url, language,
+                                      service, operation]))
+        except KeyError as error:
+            print(f"ERROR: example missing a required key: {example}.")
```

Comment:
I'm assuming this is meant to be a catch-all for all the dict key calls in this code block/nested for-loops. It might be useful to know which key is specifically causing you issues when this exception is thrown.

Reply:
Good point, I added this to the output.